### ISSUE TITLE
SDPA decomposition does not preserve intermediate dtype for fp16/bf16 inputs

### ISSUE THREAD

**Issue by oscarandersson8218** [pytorch/executorch#22012]
When exporting an fp16/bf16 model, `aten.scaled_dot_product_attention` is decomposed during `to_edge` before backend partitioning. For fp16/bf16 inputs, the current decomposition promotes SDPA intermediates to fp32, then casts the final result back to the original output dtype. In other words, the public output dtype is preserved, but the intermediate dtype is not.

Is this fp32 intermediate behavior intended to be unconditional, or should users/backends have a way to request dtype-preserving SDPA decomposition?

More generally, what should users/backends expect from Core ATen decompositions with respect to dtype? Should decompositions generally preserve the dtype of intermediate operations when the original op is exported with fp16/bf16 inputs, or is it expected that decompositions may introduce higher-precision opmath intermediates as long as the public operator output dtype is preserved?

Minimal example:

```python
import torch
from executorch.exir import to_edge

class M(torch.nn.Module):
    def forward(self, q, k, v):
        return torch.nn.functional.scaled_dot_product_attention(
            q, k, v, dropout_p=0.0
        )

inputs = tuple(torch.randn(1, 2, 4, 8, dtype=torch.float16) for _ in range(3))
ep = torch.export.export(M(), inputs)

edge = to_edge(ep)
print(edge.exported_program().graph_module)
```

After to_edge, the graph contains fp32 intermediates from the SDPA decomposition even though the model inputs and final SDPA output are fp16. The same applies to bf16.

This is understandable from a numerical-stability perspective, especially around the score matmul and softmax. However, since SDPA is decomposed before backend partitioning, backend-specific passes do not see the original SDPA op and cannot easily choose a lower-precision SDPA implementation.

cc @digantdesai @freddan80 @per @zingo @mansnils @Sebastian-Larsson @robell @rascani

**Comment by JacobSzwejbka** [comment]
Typically what we see with sdpa is that backends choose to preserve the op in to_edge_transform_and_lower and then lower it directly instead of its decomposition. On cpu we prefer to do an eager module transformation to use a custom op that handles sdpa.

**Comment by JacobSzwejbka** [comment]
Is the behavior of promoting to fp32 internally for compute different then what ATen does for fp16 sdpa?

**Comment by oscarandersson8218** [comment]
> Typically what we see with sdpa is that backends choose to preserve the op in to_edge_transform_and_lower and then lower it directly instead of its decomposition. On cpu we prefer to do an eager module transformation to use a custom op that handles sdpa.

I see, thanks. So one option for us would be to preserve the SDPA op through `to_edge_transform_and_lower` and then decompose/lower it in the backend instead.

> Is the behavior of promoting to fp32 internally for compute different then what ATen does for fp16 sdpa?

I'm not 100% sure, but I suspect they are aligned. If ATen fp16/bf16 SDPA computes internally in fp32, then I agree the current decomposition is a reasonable default from a semantics/numerics perspective.

The concern I am trying to clarify is more about user/backend control. If an end user exports a model in fp16/bf16, I think it is reasonable for them to expect that as much as possible of the model can be executed in that dtype, provided they are willing to accept the numerical tradeoff. From what I can tell, a dtype-preserving SDPA decomposition/lowering should be possible, even if it is not the default.

So perhaps the right conclusion is that the default Core ATen decomposition should keep matching ATen behavior, while backends that want a different precision tradeoff should preserve SDPA and lower it directly. Does that match your expectation @JacobSzwejbka ?

**Comment by oscarandersson8218** [comment]
@JacobSzwejbka any thoughts on my comment above?
